### PR TITLE
WT-7229 Align out of order and mixed mode handling

### PR DESCRIPTION
**Pull request by quchenhao**

### CODE REVIEW

**Comment by quchenhao** [comment]
@bvpvamsikrishna @hbokhari This is ready for review.

**Comment by quchenhao** [comment]
@hbokhari @bvpvamsikrishna Can you look at this now?

File: src/include/btmem.h [hunk 1]
```diff
@@ -1187,17 +1186,17 @@ struct __wt_update_value {
  * avoid heap allocation, add a few additional slots to that array.
  */
 #define WT_MAX_MODIFY_UPDATE 10
-#define WT_MODIFY_VECTOR_STACK_SIZE (WT_MAX_MODIFY_UPDATE + 10)
+#define WT_UPDATE_VECTOR_STACK_SIZE 20
```
Comment (hbokhari) [inline src/include/btmem.h:1189]: Curious why we are increasing the stack size?

File: src/history/hs_rec.c [hunk 2]
```diff
@@ -680,16 +646,25 @@ __wt_hs_insert_updates(
...
- * __wt_hs_delete_key_from_ts --
- *     Delete history store content of a given key from a timestamp.
+ * __wt_hs_delete_reinsert_key_from_ts --
+ *     Delete history store content of a given key from a timestamp and optionally reinsert them
+ *     with ts-1 timestamp.
  */
 int
-__wt_hs_delete_key_from_ts(WT_SESSION_IMPL *session, WT_CURSOR *hs_cursor, uint32_t btree_id,
-  const WT_ITEM *key, wt_timestamp_t ts, bool reinsert)
+__wt_hs_delete_reinsert_key_from_ts(WT_SESSION_IMPL *session, WT_CURSOR *hs_cursor,
```
Comment (hbokhari) [inline src/history/hs_rec.c:654]: I think there is no need for word `reinsert` in the name. I believe the function arguments are self explanatory.

File: src/history/hs_rec.c [hunk 3]
```diff
@@ -711,14 +690,14 @@ __wt_hs_delete_key_from_ts(WT_SESSION_IMPL *session, WT_CURSOR *hs_cursor, uint3
...
- *     reinserting them at the timestamp that is currently being inserted.
+ * __hs_delete_reinsert_from_pos --
+ *     Delete updates in the history store if they are larger or equal to the specified timestamp
+ *     and reinsert them with ts-1 timestamp. This function works by looking ahead of the current
+ *     cursor position for entries for the same key, removing them.
  */
 static int
-__hs_fixup_out_of_order_from_pos(WT_SESSION_IMPL *session, WT_CURSOR *hs_cursor, WT_BTREE *btree,
-  const WT_ITEM *key, wt_timestamp_t ts, uint64_t *counter)
+__hs_delete_reinsert_from_pos(WT_SESSION_IMPL *session, WT_CURSOR *hs_cursor, uint32_t btree_id,
```
Comment (hbokhari) [inline src/history/hs_rec.c:699]: Again no need for word `reinsert` in the function name I believe.

File: src/history/hs_rec.c [hunk 4]
```diff
@@ -680,16 +646,25 @@ __wt_hs_insert_updates(
...
+    uint32_t hs_btree_id;
     bool hs_read_committed;
 
+    /* If we will delete all the updates of the key from the history store, we should not reinsert
+     * any update. */
+    WT_ASSERT(session, ts > WT_TS_NONE || !reinsert);
+
     hs_read_committed = F_ISSET(hs_cursor, WT_CURSTD_HS_READ_COMMITTED);
     if (!hs_read_committed)
         F_SET(hs_cursor, WT_CURSTD_HS_READ_COMMITTED);
```
Comment (hbokhari) [inline src/history/hs_rec.c:670]: Do we also need to use the flag `WT_CURSTD_HS_READ_ALL` for this search_near scenario?

File: src/history/hs_rec.c [hunk 5]
```diff
@@ -160,29 +158,25 @@ __hs_insert_record(WT_SESSION_IMPL *session, WT_CURSOR *cursor, WT_BTREE *btree,
...
-            if (!hs_read_all_flag)
-                F_CLR(cursor, WT_CURSTD_HS_READ_ALL);
-        }
-        if (ret == 0)
-            WT_ERR(__hs_fixup_out_of_order_from_pos(
-              session, cursor, btree, key, tw->start_ts, &counter));
+        if (!hs_read_all_flag)
+            F_CLR(cursor, WT_CURSTD_HS_READ_ALL);
+    }
+    if (ret == 0) {
```
Comment (hbokhari) [inline src/history/hs_rec.c:177]: Do we need curly braces here?

File: src/history/hs_rec.c [hunk 6]
```diff
@@ -680,16 +646,25 @@ __wt_hs_insert_updates(
 }
 
 /*
- * __wt_hs_delete_key_from_ts --
- *     Delete history store content of a given key from a timestamp.
+ * __wt_hs_delete_reinsert_key_from_ts --
+ *     Delete history store content of a given key from a timestamp and optionally reinsert them
+ *     with ts-1 timestamp.
```
Comment (hbokhari) [inline src/history/hs_rec.c:651]: I believe we are bounding the scope of this function by only reinserting records at ts-1. Can we explicitly provide the reinsert timestamp to the function?

File: src/history/hs_rec.c [hunk 7]
```diff
@@ -160,29 +158,25 @@ __hs_insert_record(WT_SESSION_IMPL *session, WT_CURSOR *cursor, WT_BTREE *btree,
...
-        /*
-         * If there were no keys equal to or less than our target key, we would have received
-         * WT_NOTFOUND. In that case we need to search again with a higher timestamp as the cursor
-         * would not be positioned correctly.
-         */
-        if (ret == 0)
-            WT_ERR_NOTFOUND_OK(cursor->next(cursor), true);
-        else {
-            F_SET(cursor, WT_CURSTD_HS_READ_ALL);
+    if (ret == 0)
```
Comment (hbokhari) [inline src/history/hs_rec.c:166]: Just noticed that we are doing another search_near() in `__hs_delete_reinsert_from_pos`. Can we somehow simplify the logic and here probably get rid of next() or search_near() call here?

Comment (hbokhari) [reply src/history/hs_rec.c:166]: Actually scratch that - I mixed up the function with __wt_hs_delete_reinsert_key_from_ts()

File: src/history/hs_rec.c [hunk 8]
```diff
@@ -711,14 +690,14 @@ __wt_hs_delete_key_from_ts(WT_SESSION_IMPL *session, WT_CURSOR *hs_cursor, uint3
 }
 
 /*
- * __hs_fixup_out_of_order_from_pos --
- *     Fixup existing out-of-order updates in the history store. This function works by looking
- *     ahead of the current cursor position for entries for the same key, removing them and
- *     reinserting them at the timestamp that is currently being inserted.
+ * __hs_delete_reinsert_from_pos --
+ *     Delete updates in the history store if they are larger or equal to the specified timestamp
+ *     and reinsert them with ts-1 timestamp. This function works by looking ahead of the current
```
Comment (hbokhari) [inline src/history/hs_rec.c:695]: Again the idea of reinserting at ts-1 is bit confusing. Maybe we should provide the new timestamp explicitly.

File: src/history/hs_rec.c [hunk 9]
```diff
@@ -160,29 +158,25 @@ __hs_insert_record(WT_SESSION_IMPL *session, WT_CURSOR *cursor, WT_BTREE *btree,
...
-        }
-        if (ret == 0)
-            WT_ERR(__hs_fixup_out_of_order_from_pos(
-              session, cursor, btree, key, tw->start_ts, &counter));
+        if (!hs_read_all_flag)
+            F_CLR(cursor, WT_CURSTD_HS_READ_ALL);
+    }
+    if (ret == 0) {
+        WT_ERR(__hs_delete_reinsert_from_pos(
+          session, cursor, btree->id, key, tw->start_ts + 1, true, &counter));
```
Comment (hbokhari) [inline src/history/hs_rec.c:179]: We are setting `tw->start_ts + 1` here because we actually want to delete them from `tw->start_ts`?

**Review by hbokhari — COMMENTED** [review]
These is a review for part of code I have gone through so far.

File: src/include/btmem.h (hunk 1 above)
Comment (quchenhao) [reply src/include/btmem.h:1189]: It's the same size. I just don't want the stack size to depend on the number of modies.

File: src/history/hs_rec.c (hunk 9 above)
Comment (quchenhao) [reply src/history/hs_rec.c:179]: Yes, that's correct.

File: src/history/hs_rec.c (hunk 6 above)
Comment (quchenhao) [reply src/history/hs_rec.c:651]: I don't want to give too much freedom to the caller at the moment. If such use case appear, we can do that.

File: src/history/hs_rec.c (hunk 4 above)
Comment (quchenhao) [reply src/history/hs_rec.c:670]: Yes, why you think it's not needed? This is from your change. I'm just merging it.

Comment (hbokhari) [reply src/history/hs_rec.c:669]: I didn't include the flag then. Can you review the need for flag here given we are refactoring out-of-order update code?

Comment (quchenhao) [reply src/history/hs_rec.c:669]: What do you mean you didn't include? If you look at the code. The code is in the base commit.

Comment (hbokhari) [reply src/history/hs_rec.c:669]: What I am saying is that I never analysed if we need to put flag `WT_CURSTD_HS_READ_ALL` here (before search near).

File: src/history/hs_rec.c [hunk 10]
```diff
@@ -464,54 +432,66 @@ __wt_hs_insert_updates(
...
             WT_ASSERT(session, upd->type == WT_UPDATE_STANDARD || upd->type == WT_UPDATE_MODIFY);
 
-            tw.durable_start_ts = upd->durable_ts;
-            tw.start_ts = upd->start_ts;
-            tw.start_txn = upd->txnid;
             tombstone = NULL;
-            __wt_modify_vector_peek(&modifies, &prev_upd);
+            __wt_update_vector_peek(&updates, &prev_upd);
+
+            if (out_of_order_ts_updates.size > 0) {
```
Comment (hbokhari) [inline src/history/hs_rec.c:482]: Can we make it a macro?

Comment (quchenhao) [reply src/history/hs_rec.c:482]: Why do you think a macro is better?

Comment (hbokhari) [reply src/history/hs_rec.c:482]: Just for the sake of abstraction - if we somehow change the variable names in underlying implementation of vector, we don't have to change the code. But then again it's just a styling thing.

**Comment by quchenhao** [comment]
Stress test
https://spruce.mongodb.com/version/60640941c9ec443fa8a92cec/tasks?sorts=STATUS%3AASC%3BBASE_STATUS%3ADESC

**Comment by quchenhao** [comment]
Mongodb patch build
https://spruce.mongodb.com/patch/60640a3657e85a6df3520683/configure/tasks

**Comment by quchenhao** [comment]
I don't see any failures related to the change in the patch builds

File: src/txn/txn_rollback_to_stable.c [hunk 11]
```diff
@@ -377,10 +377,14 @@ __rollback_ondisk_fixup_key(WT_SESSION_IMPL *session, WT_REF *ref, WT_PAGE *page
          * stop timestamp if the original update's commit timestamp is out of order. We may see
          * records newer than or equal to the onpage value if eviction runs concurrently with
          * checkpoint. In that case, don't verify the first record.
+         *
+         * If we have fixed the out-of-order timestamp, then the newer update which has been
+         * reinserted with the fixed timestamp may have a durable timestamp that is smaller than the
```
Comment (hbokhari) [inline src/txn/txn_rollback_to_stable.c:382]: If fixed timestamp is always an older timestamp then can we use the word "reinserted with the an older timestamp"

File: src/history/hs_rec.c (hunk 4 above)
Comment (quchenhao) [reply src/history/hs_rec.c:669]: I believe we should here.

File: src/history/hs_rec.c (hunk 8 above)
Comment (hbokhari) [reply src/history/hs_rec.c:694]: Delete updates in the history store if they are -> Delete updates in the history store if the start timestamp of the update is.

reinsert them with ts-1 timestamp -> and optionally reinsert them with ts-1 timestamp

File: src/history/hs_rec.c [hunk 12]
```diff
@@ -741,32 +719,29 @@ __hs_fixup_out_of_order_from_pos(WT_SESSION_IMPL *session, WT_CURSOR *hs_cursor,
...
+        if (hs_ts >= ts)
             break;
     }
     if (ret == WT_NOTFOUND)
         return (0);
     WT_ERR(ret);
 
     /*
-     * The goal of this fixup function is to move out-of-order content to maintain ordering in the
+     * The goal of this function is to move out-of-order content to maintain ordering in the
```
Comment (hbokhari) [inline src/history/hs_rec.c:744]: Do we want to add an example of how we deal with mixed mode timestamps here?

Comment (quchenhao) [reply src/history/hs_rec.c:744]: Added

File: src/history/hs_rec.c [hunk 13]
```diff
@@ -296,7 +288,17 @@ __wt_hs_insert_updates(
     WT_RET(__wt_curhs_open(session, NULL, &hs_cursor));
     F_SET(hs_cursor, WT_CURSTD_HS_READ_COMMITTED);
 
-    __wt_modify_vector_init(session, &modifies);
+    __wt_update_vector_init(session, &updates);
+    /*
+     * We use another stack to store the out-of-order timestamp updates (including updates without a
+     * timestamp). We walk the update chain from the newest to the oldest. Once an out-of-order
+     * timestamp update is detected, if it has lower timestamp than the head of the stack, it is
```
Comment (hbokhari) [inline src/history/hs_rec.c:295]: is detected, if it has lower timestamp than the head ->is detected, and it has a lower timestamp than the head

File: src/history/hs_rec.c [hunk 14]
```diff
@@ -741,32 +729,29 @@ __hs_fixup_out_of_order_from_pos(WT_SESSION_IMPL *session, WT_CURSOR *hs_cursor,
...
-     * Position ourselves at the beginning of the key range that we may have to fixup. Prior to
-     * getting here, we've positioned our cursor at the end of a key/timestamp range and then done a
-     * "next". Normally that would leave us pointing at higher timestamps for the same key (if any)
-     * but in the case where our insertion timestamp is the lowest for that key, our cursor may be
-     * pointing at the previous key and can potentially race with additional key insertions. We need
-     * to keep doing "next" until we've got a key greater than the one we attempted to position
-     * ourselves with.
-     */
+
+    /* If we will delete all the updates of the key from the history store, we should not reinsert
```
Comment (hbokhari) [inline src/history/hs_rec.c:733]: Reformat the multi-line comment here.

File: src/history/hs_rec.c [hunk 15]
```diff
@@ -784,182 +769,95 @@ __hs_fixup_out_of_order_from_pos(WT_SESSION_IMPL *session, WT_CURSOR *hs_cursor,
...
         WT_ASSERT(session, cmp == 0);
 #endif
         /*
          * If we got here, we've got out-of-order updates in the history store.
          *
-         * Our strategy to rectify this is to remove all records for the same key with a higher
-         * timestamp than the one that we're inserting on and reinsert them at the same timestamp
-         * that we're inserting with.
+         * Our strategy to rectify this is to remove all records for the same key with a higher or
+         * equal timestamp to the specified timestamp and reinsert them at the smaller timestamp,
```
Comment (hbokhari) [inline src/history/hs_rec.c:798]: a higher or equal timestamp to the -> a timestamp equal or higher than the

File: src/history/hs_rec.c [hunk 16]
```diff
@@ -784,182 +769,95 @@ __hs_fixup_out_of_order_from_pos(WT_SESSION_IMPL *session, WT_CURSOR *hs_cursor,
      * 2     foo 3  1       bbb
      * 2     foo 3  2       ccc
      * 2     foo 3  3       ddd
+     *
+     * * Another example, if we're inserting an update at timestamp 0 with value ddd:
```
Comment (hbokhari) [inline src/history/hs_rec.c:773]: Remove extra asterisk before "Another example".

File: src/history/hs_rec.c [hunk 17]
```diff
@@ -784,182 +769,95 @@ __hs_fixup_out_of_order_from_pos(WT_SESSION_IMPL *session, WT_CURSOR *hs_cursor,
...
+              __wt_timestamp_to_string(hs_cbt->upd_value->tw.durable_stop_ts, ts_string[3]),
+              __wt_timestamp_to_string(ts, ts_string[4]));
+
+            hs_insert_tw.start_ts = hs_insert_tw.durable_start_ts = ts - 1;
             hs_insert_tw.start_txn = hs_cbt->upd_value->tw.start_txn;
 
-            hs_insert_tw.stop_ts = hs_insert_tw.durable_stop_ts = WT_TS_NONE;
+            /*
+             * We're going to be inserting something immediately after with the smaller timestamp.
+             * Either another moved update OR the update itself that triggered the correction. In
```
Comment (hbokhari) [inline src/history/hs_rec.c:835]: Not sure but this sentence doesn't look right.

File: src/history/hs_rec.c (hunk 4 above)
Comment (hbokhari) [reply src/history/hs_rec.c:678]: @quchenhao have you tried setting `WT_CURSTD_HS_READ_ALL` here?

**Review by hbokhari — APPROVED** [review]
I have left few minor comments but otherwise this change is looking good now.

File: src/history/hs_rec.c (hunk 4 above)
Comment (quchenhao) [reply src/history/hs_rec.c:678]: No. I was mistakenly think you are talking about `WT_CURSTD_HS_READ_COMMITTED` flag.

Comment (quchenhao) [reply src/history/hs_rec.c:678]: @hbokhari I checked the old code and I think we need to set the flag here. However, I don't want to do this in this ticket. Can you create a separate ticket for this task?

**Comment by quchenhao** [comment]
@hbokhari Can you create a separate ticket to set the flag? This ticket has already got too much.

File: src/support/update_vector.c [hunk 18]
```diff
@@ -0,0 +1,111 @@
...
+        if (updates->allocated_bytes == 0 && updates->size == WT_UPDATE_VECTOR_STACK_SIZE) {
+            migrate_from_stack = true;
+            updates->listp = NULL;
+        }
+        WT_ERR(__wt_realloc_def(
+          updates->session, &updates->allocated_bytes, updates->size + 1, &updates->listp));
+        if (migrate_from_stack)
+            memcpy(updates->listp, updates->list, sizeof(updates->list));
+    }
+    updates->listp[updates->size++] = upd;
```
Comment (bvpvamsikrishna) [inline src/support/update_vector.c:46]: Please don't use '++' notation in array referencing, prone to mistakes.

File: src/support/update_vector.c [hunk 19]
```diff
@@ -0,0 +1,111 @@
...
+/*
+ * __wt_update_vector_pop --
+ *     Pop an update pointer off a update vector.
+ */
+void
+__wt_update_vector_pop(WT_UPDATE_VECTOR *updates, WT_UPDATE **updp)
+{
+    WT_ASSERT(updates->session, updates->size > 0);
+
+    *updp = updates->listp[--updates->size];
```
Comment (bvpvamsikrishna) [inline src/support/update_vector.c:75]: Please don't use '--' notation in array referencing, prone to mistakes.

File: src/history/hs_rec.c [hunk 20]
```diff
@@ -742,32 +732,29 @@ __hs_fixup_out_of_order_from_pos(WT_SESSION_IMPL *session, WT_CURSOR *hs_cursor,
...
         WT_ERR(hs_cursor->get_key(hs_cursor, &hs_btree_id, &hs_key, &hs_ts, &hs_counter));
-        WT_ASSERT(session, hs_btree_id == btree->id);
+        WT_ASSERT(session, hs_btree_id == btree_id);
 #ifdef HAVE_DIAGNOSTIC
         WT_ERR(__wt_compare(session, NULL, &hs_key, key, &cmp));
         WT_ASSERT(session, cmp == 0);
 #endif
-        if (hs_ts > ts)
+        /* We find a key that is larger or equal to the specified timestamp*/
+        if (hs_ts >= ts)
```
Comment (bvpvamsikrishna) [inline src/history/hs_rec.c:749]: Why do we need an entry with the same timestamp?

**Comment by bvpvamsikrishna** [comment]
@quchenhao, we will discuss this PR online on monday.  Meanwhile could you trigger mongodb correctness patch and perf build and post the link in jira ticket.

**Review by bvpvamsikrishna — APPROVED** [review]
lgtm, if there is no fallout from patch build.

**Pull request merged**